### PR TITLE
test: qualify paged scheduler parity

### LINKED ISSUE #1773: fix(inference): server and CLI diverge by one token on LLM-jp-4-vl-9B
## Problem / Background

On `llm-jp/llm-jp-4-vl-9b-beta` (local `models/mlx/llm-jp-4-vl-9b-beta`), `mlxcel-server` answers `この画像は、単色の赤色で塗りつぶされた正方形のキャンバスです。` where `mlxcel generate` and the checkpoint's own `modeling_llmjpvl.py` both answer `...正方形の図形です。` One token differs, and it reproduced stably across three requests, so it is a path difference and not run-to-run noise. Found while implementing #1362 (merged as PR #1737, commit `13a0df44`), attributed at the time but deliberately not fixed and never filed; the evidence is `TECHNICAL_REPORTS/1737-llmjpvl-siglip2-pixel-shuffle-llama-qwen3-20260910.en.md` section 4.3.

Attribution is settled and this issue does not ask to redo it. The same server started with `--max-batch-size 1` returns `図形` and matches the CLI and the reference exactly, and the prompt is pinned rather than argued: `one_image_prompts_are_token_exact_against_the_reference_processor` (`src/multimodal/llmjp_vl_prompt_parity_tests.rs:166`) compares every assembled prompt id against the reference processor's `input_ids`, and `the_server_image_request_renders_the_same_prompt_as_the_cli` (`src/server/llmjp_chat_template_tests.rs:109`) asserts byte equality between the two front ends' renders. What is open is whether a batched-path token difference against the unbatched path is a contract violation here or expected behavior nobody wrote down.

## Current Behavior

Two written statements in this tree collide on that question, and neither is reachable by an operator.

- `docs/CONTINUOUS_BATCHING.md:252` and `docs/turbo-kv-cache.md:423` both assert "Paged decode is byte-identical to the dense backend", citing `tests/paged_scheduler_parity.rs` (RMS 0). That test covers qwen3 and llama3 fp16 at B=1 and B=2, is `#[ignore]`, and its documented run command is `--features metal,accelerate`. It has no CUDA arm and no VLM arm.
- `src/server/batch/scheduler_cohort_parity_tests.rs:30-38` and `src/server/batch/scheduler_seed_determinism_tests.rs:27-32` say the opposite for the batched-versus-B=1 boundary, name it "the documented #203 / #325 / #326 jitter class, not a correctness bug", and deliberately refuse to assert byte identity across it. `docs/benchmark_results/mlx-pin-upgrade-2026-07-03.md:90-100` records that since MLX #3704 batched-versus-B=1 hidden states are not bitwise identical on CUDA at all.

One detail narrows the search and should not be skipped: `--max-batch-size 1` does more than narrow the decode width. `effective_decode_storage_backend` (`src/server/batch/scheduler/mod.rs:216-229`) computes `paged_available = max_batch_size > 1 && supports_batching && supports_paged_decode_backend`, so at `1` the `Auto` default resolves to `Dense` instead of `Paged`. With a single in-flight request the decode width is 1 either way, because `dispatch_sync_decode` (`src/server/batch/scheduler/decode_tick.rs:269-276`) routes `seq_ids.len() <= 1` to `decode_single_step`. llmjpvl delegates `supports_batching` and `supports_paged_decode_backend` to its Llama backbone (`src/vision/llmjp_vl.rs:338-352`), so the family is pool-backed and paged-eligible.

## Proposed Solution

1. Bisect on the same checkpoint before writing anything: run the server at the default `--max-batch-size 4` plus `--decode-storage-backend dense`. If that returns `図形`, the paged storage backend is the variable and the two "byte-identical to the dense backend" claims are wrong or under-qualified on CUDA. If it still returns `キャンバス`, the backend is not the variable, and the remaining candidates are the decode lookahead pipeline (`run_decode_tick`, `src/server/batch/scheduler/decode_tick.rs:288`) and the scheduler decode step itself against the CLI generate loop.
2. State the contract in `docs/CONTINUOUS_BATCHING.md`, in a short subsection under "Continuous batching": whether `mlxcel-server` at shipped defaults is expected to be token-exact against `mlxcel generate` for the same prompt at `temperature 0`. If it is not, say so plainly and name `--max-batch-size 1` and `--decode-storage-backend dense` as the settings that buy it back. Qualify the paged claim at `docs/CONTINUOUS_BATCHING.md:252` and `docs/turbo-kv-cache.md:423` with the backend, dtype, and families it was actually measured on.
3. Pin the intended contract with a test, not the incidental output. Do not assert `図形` or any specific Japanese string. The property worth pinning is paged-versus-dense equivalence on the backend where the divergence was seen, so extend `tests/paged_scheduler_parity.rs` with an arm runnable under `--features cuda` on an existing small llama3 or qwen3 checkpoint, asserting whatever the step 1 bisect establishes: byte identity if it holds, or a stated tolerance plus greedy-token equality if it does not. Whichever way it lands, the test must fail if the answer changes.

Rejected: fixing the token. Chasing bit-exactness in one family's near-tie is not tractable here and the port itself is validated token-exact on the CLI against the checkpoint's own reference.

## Scope

**In scope:** the bisect above; `docs/CONTINUOUS_BATCHING.md`; `docs/turbo-kv-cache.md:423`; a CUDA-runnable arm in `tests/paged_scheduler_parity.rs`.

**Out of scope:** changing any kernel, reduction order, or MLX pin; making llmjpvl's server output match its CLI output token-for-token; the XLA/IREE operator contracts tracked in #932.

## Implementation Notes

- **Reuse**: `tests/paged_scheduler_parity.rs` already drives `CachePool::allocate_with_layout` against `CachePool::allocate` and explains why it uses `CachePool` rather than `BatchScheduler` (which is `pub(crate)`). Follow that structure and its soft-skip-when-the-model-is-absent convention rather than adding a new harness.
- **Constraints**: the divergence was observed on Linux aarch64 / GB10 / CUDA sm_121. A Metal-only run cannot answer this issue. Set `MLX_ENABLE_TF32=0` for any numeric comparison, since MLX defaults it on and a 10-bit mantissa is not evidence about f32 reduction order.
- **Edge cases**: a lone request under 4096 visible KV tokens does not reach the fused paged v2 kernel at all (dispatch floor documented at `docs/CONTINUOUS_BATCHING.md:270-284`), and this prompt is 297 tokens, so the gather-then-SDPA reference path is what ran. Say which path the bisect exercised.
- **Error handling**: no runtime behavior change is expected. If the bisect shows a real correctness bug rather than reduction-order drift, stop and file a separate issue rather than widening this one.

## Acceptance Criteria

- [ ] The bisect result is recorded in the issue: which of `--max-batch-size 1` and `--decode-storage-backend dense` restores the CLI answer, on CUDA, with the exact commands and outputs.
- [ ] `docs/CONTINUOUS_BATCHING.md` states whether server defaults are token-exact against `mlxcel generate` at `temperature 0`, and names the settings that make them so when they are not.
- [ ] The unqualified "byte-identical to the dense backend" wording at `docs/CONTINUOUS_BATCHING.md:252` and `docs/turbo-kv-cache.md:423` is either substantiated on CUDA or narrowed to what was measured.
- [ ] `tests/paged_scheduler_parity.rs` has an arm that runs under `--features cuda` and asserts the contract decided above; it fails when the contract is broken.
- [ ] No test asserts a specific llmjpvl output string.

## Verification

```
./target/release/mlxcel generate -m models/mlx/llm-jp-4-vl-9b-beta \
    --image solid_red.png -p "この画像について説明してください。" -n 32 --temp 0 --profile

./target/release/mlxcel-server -m models/mlx/llm-jp-4-vl-9b-beta --port 8080
./target/release/mlxcel-server -m models/mlx/llm-jp-4-vl-9b-beta --port 8080 --max-batch-size 1
./target/release/mlxcel-server -m models/mlx/llm-jp-4-vl-9b-beta --port 8080 --decode-storage-backend dense

curl -s localhost:8080/v1/chat/completions -H 'content-type: application/json' \
  -d '{"model":"llm-jp-4-vl-9b-beta","temperature":0,"max_tokens":32,"messages":[{"role":"user","content":[{"type":"text","text":"この画像について説明してください。"},{"type":"image_url","image_url":{"url":"data:image/png;base64,<solid red 512x512>"}}]}]}'

MLX_ENABLE_TF32=0 cargo test --test paged_scheduler_parity --release --features cuda \
    -- --ignored --nocapture --test-threads=1
cargo clippy -p mlxcel --lib --tests --features cuda -- -D warnings
cargo fmt --all -- --check
```

A pass is: the three server invocations produce a recorded, reproducible answer each; the new parity arm runs (not skipped) on CUDA and passes; the docs no longer leave the question unanswered.

## Technical Considerations

The port itself is not suspect. Both backbones reproduce the reference processor's prompt token count and its greedy output exactly on the CLI, and Jagle-VL is byte-identical to the reference on the server as well. Only the 9B, only on the server, only at one token. Related: #932 (operator-level numeric contracts, `status:blocked`), #1558 and #1563 (the same "recorded, not gated" reproducibility class on sm_70).

### PR DESCRIPTION
**Pull request by inureyes**
## Summary

- Narrow the paged decode parity documentation to the scheduler-shaped qwen3/llama3 evidence the integration test actually exercises instead of claiming universal byte identity.
- State that continuous-batching defaults are a serving-throughput contract rather than a universal token-exactness guarantee against `mlxcel generate`, and document `--decode-storage-backend dense` as the first backend-only bisect before `--max-batch-size 1`.
- Add a CUDA-gated qwen3 parity arm that checks greedy-token equality plus bounded full-vocabulary selector-logit RMS with TF32 disabled, without asserting an LLM-jp Japanese output string.

## CUDA bisect

On spark-101 (NVIDIA GB10, compute capability 12.1, driver 580.173.02, Linux aarch64), build `2967db004ed0e34fdefde879abd6ed00427df45c` no longer reproduced the original `llm-jp/llm-jp-4-vl-9b-beta` default-server token difference: CLI, default-batch server with `--decode-storage-backend dense`, default server, and server with `--max-batch-size 1` all returned `この画像は、単色の赤色で塗りつぶされた正方形の図形です。` with `prompt_tokens=297`, `completion_tokens=15`, and `MLX_ENABLE_TF32=0`. The default server reported `decode_storage=auto`. Because the prompt is 297 tokens, this exercises gather-then-SDPA below the fused paged-v2 dispatch floor; it is not fused paged-v2 evidence. The full command/output transcript is recorded on #1773.

## Changes during review

- Align each trace row with the logits that selected its emitted token, including the terminal prefill row for the first token.
- Reject empty and non-finite selector-logit rows before computing relative RMS.
- Document `MLXCEL_REQUIRE_MODELS=1` for evidence runs so a missing checkpoint cannot soft-skip unnoticed.
- Correct the batched-path commentary for the current whole-batch pooled decode helper and clarify that dense/single-request controls are diagnostic rather than universal exactness guarantees.

## Test plan

- [x] `cargo fmt --all -- --check`
- [x] `git diff --check origin/main...HEAD`
- [x] `cargo test --test paged_scheduler_parity --profile test-fast --features metal,accelerate --no-run`
- [x] `cargo clippy -p mlxcel --test paged_scheduler_parity --features metal,accelerate -- -D warnings`
- [x] `MLX_ENABLE_TF32=0 cargo test --test paged_scheduler_parity --release --features cuda -- --ignored --nocapture --test-threads=1` on spark-101 at `363d664849603e13dcbeb402ba14bfd1e7d23e84`; 5 passed, qwen3 executed, and llama3 soft-skipped because that checkpoint was absent.
- [x] `MLXCEL_REQUIRE_MODELS=1 MLX_ENABLE_TF32=0 cargo test --test paged_scheduler_parity --release --features cuda cuda_paged_scheduler_qwen3_matches_dense_within_logit_contract -- --ignored --nocapture` on spark-101 at final head `2dc42d9db33072a45fe26c36849eefacfbcdccee`; the checkpoint was required and present, dense/paged greedy traces were identical, and the contract passed.
- [x] `cargo clippy -p mlxcel --lib --tests --features cuda -- -D warnings` on spark-101 at `2967db004ed0e34fdefde879abd6ed00427df45c`.

Closes #1773

### CODE REVIEW

**Comment by inureyes** [comment]
## Implementation Review Summary

### Intent
> Qualify the paged/dense decode contract for #1773, record the GB10 bisect honestly, and add a CUDA-runnable parity gate without changing runtime inference behavior.

### Findings Addressed
- [x] Aligned the RMS trace with the logits that actually selected each emitted token, including the prefill terminal row instead of only post-token next rows. (MEDIUM)
- [x] Updated the test commentary to describe the current #899 whole-batch pooled paged decode/gather fallback path instead of the older per-sequence fallback wording. (LOW)
- [x] Clarified that `--max-batch-size 1` and dense decode are diagnostic/oracle controls, not an unmeasured token-exactness guarantee for every model family. (LOW)

### Remaining Items
- [ ] Real CUDA runtime validation has not been rerun after review follow-up `363d6648`; the recorded GB10 bisect and CUDA test evidence remains from `2967db004ed0e34fdefde879abd6ed00427df45c`. (MEDIUM)

### Verification
- [x] Requirements from the issue are represented in docs/tests and the #1773 transcript records the all-`図形` GB10 result at the measured commit.
- [x] No llmjpvl Japanese output fixture was added.
- [x] The CUDA arm is feature-gated and soft-skips only when CUDA or the small checkpoint is absent.
- [x] Local `cargo fmt --all -- --check`, `cargo test --test paged_scheduler_parity --profile test-fast --features metal,accelerate --no-run`, and `cargo clippy -p mlxcel --test paged_scheduler_parity --features metal,accelerate -- -D warnings` passed.
- [ ] Final-head real CUDA runtime test pass is pending for `363d6648`.

**Comment by inureyes** [comment]
## Security and performance review summary

- No CRITICAL or HIGH security/performance findings found in the docs plus ignored parity-test diff.
- MEDIUM fixed in `2dc42d9d`: the CUDA evidence command now documents `MLXCEL_REQUIRE_MODELS=1`, so a missing qwen3 checkpoint cannot be mistaken for a real contract pass when final evidence is required.
- LOW fixed in `2dc42d9d`: the relative-RMS helper now rejects empty and non-finite logit rows explicitly before computing the ratio, instead of depending on a later tolerance comparison to make NaN/Inf fail.
- Reviewed host-memory amplification from copying full-vocabulary selector logits per compared step; it is bounded by the ignored-test constants and the two small checkpoints, so I did not change the storage shape in this pass.

Verification: `cargo fmt --all -- --check`; `cargo test --test paged_scheduler_parity --profile test-fast --features metal,accelerate --no-run`; `cargo clippy -p mlxcel --test paged_scheduler_parity --features metal,accelerate -- -D warnings`.

Unresolved: I did not rerun the real CUDA runtime suite after `2dc42d9d`; the qwen3 CUDA evidence handed to me was from `363d6648` on spark-101, where the exact suite passed 5/5 with qwen3 running and llama3 soft-skipping.

**Comment by inureyes** [comment]
## Finalization update

- The finalization pass found no additional correctness, documentation, test-coverage, or repository-convention gaps at `2dc42d9db33072a45fe26c36849eefacfbcdccee`; no further code changes were needed.
- Local finalization checks passed: `cargo fmt --all -- --check`, `git diff --check origin/main...HEAD`, `cargo test --test paged_scheduler_parity --profile test-fast --features metal,accelerate --no-run`, and `cargo clippy -p mlxcel --test paged_scheduler_parity --features metal,accelerate -- -D warnings`.
- Final-head GB10 CUDA evidence passed with model-presence enforcement and TF32 disabled: `MLXCEL_REQUIRE_MODELS=1 MLX_ENABLE_TF32=0 cargo test --test paged_scheduler_parity --release --features cuda cuda_paged_scheduler_qwen3_matches_dense_within_logit_contract -- --ignored --nocapture`.
- The CUDA test ran on spark-101 (NVIDIA GB10, compute capability 12.1, driver 580.173.02), required `/home/inureyes/models/mlx/qwen3-0.6b-4bit`, produced identical dense/paged greedy traces `[1079, 264, 5458, 304, 279, 220, 16, 15, 339, 11972, 13, 358, 614, 311, 3270, 264]`, and finished with 1 passed, 0 failed.
